### PR TITLE
MAINT: Improve implementation details and doc in `stats.shapiro`

### LINKED ISSUE #685: Statistics Review: shapiro (Trac #158)
_Original ticket http://projects.scipy.org/scipy/ticket/158 on 2006-04-02 by @rkern, assigned to @rkern._

The function `shapiro` in file source:trunk/Lib/stats/morestats.py needs review.

Please look over the StatisticsReview guidelines and add your comments below.

### PR DESCRIPTION
**Pull request by aeklant**
closes gh-685

### CODE REVIEW

**Comment by aeklant** [comment]
Please review.

There are two items pending:
1. nan input: This one is part of `stats.morestats`(i.e. it doesn't have a masked counterpart). Here is one example of what happens with nan input:
```
In [9]: x = np.arange(10.)

In [10]: x[9] = np.nan

In [11]: stats.shapiro(x)
Out[11]: (nan, 1.0)
```

2. `FutureWarning` for `a` and `reta`: If I understood correctly, the implementation should look something like this:
```
if reta or a is not None:
    warnings.warn("input parameters 'a' and 'reta' are scheduled to be removed in version 0.18.0", 
                  Future Warning)
```
I will check the tests to see if any of them need to ignore warnings to avoid failing regression tests in development versions.

Should something be added to the release notes as well?

**Comment by rgommers** [comment]
It should be ``if reta is not None or a is not None:`` but other than that, yes that's how the FutureWarning should look.

**Comment by rgommers** [comment]
Can you add the tests you did against R as regression tests, with a comment that those were verified against R?

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -1222,12 +1222,33 @@ def shapiro(x, a=None, reta=False):
...
+    >>> x = stats.norm.rvs(loc=5, scale=3, size=100)
+    >>> stats.shapiro(x)
+    (0.9772805571556091, 0.08144091814756393)
 
     """
+    x = np.ravel(x)
+    if x.dtype == object:
+        # np.ravel and np.ndarray.flatten don't work with dtype=object
+        # i.e. they will fail to flatten if there are different sized arrays.
+        x = np.hstack(x)
```
Comment (rgommers) [inline scipy/stats/morestats.py]: Why did you add this? We don't do this for any other function - users should just not use object arrays with unequal length arrays inside imho.

File: scipy/stats/morestats.py [hunk 2]
```diff
@@ -1241,7 +1262,8 @@ def shapiro(x, a=None, reta=False):
     y = sort(x)
     a, w, pw, ifault = statlib.swilk(y, a[:N//2], init)
     if ifault not in [0, 2]:
-        warnings.warn(str(ifault))
+        warnings.warn("There has been a problem in the computation. The "
```
Comment (rgommers) [inline scipy/stats/morestats.py]: Maybe add ``in `shapiro` ``, so users can more easily locate the issue in their own code.

**Comment by rgommers** [comment]
Also need the test for 2-D input.

**Comment by rgommers** [comment]
The returned result for input with nans looks correct to me. Leave as is but add a test for it I'd say.

**Comment by rgommers** [comment]
A ``Notes`` section which says that the algorithm used is from ref [4] and that for N>5000 the p-value may not be accurate (but the W test statistic will still be) would be good. And that the censoring parameters described in [4] are not used.

And a second note about power of the test: rejection of the null hypothesis when it is true is close to 5% more or less independent of sample size.

**Comment by rgommers** [comment]
@aeklant I'm done with giving new suggestions now.

File: scipy/stats/morestats.py (hunk 1 above)
Comment (aeklant) [inline scipy/stats/morestats.py]: I put this in because I added a test with different size arrays and if failed due to an unforeseen side effect.

If we are to remove this, maybe we should consider checking to make sure all input arrays are the same size and raising a ValueError if not, to let the user know.

Comment (rgommers) [inline scipy/stats/morestats.py]: We're going to remove and don't really care about ``a`` here, so there is only one input array: ``x``. So I don't really understand your comment. ``np.ravel(x)`` should be enough. If not, can you provide a more concrete example?

Comment (aeklant) [inline scipy/stats/morestats.py]: It doesn't actually make it silently return wrong values like I had thought, it just errors. If this is considered an acceptable result I can remove the added code.

e.g.
```
stats.shapiro([[1, 2], [3, 4], [5]])
```

With the above code:
```
(0.9867621660232544, 0.9671739339828491)
```
(Same as `stats.shapiro([1, 2, 3, 4, 5])` which is correct)

Without the aboove code:
```
ValueError: setting an array element with a sequence.
```

**Comment by aeklant** [comment]
Please review. If everything is well then there would only be the matter of the added check for different shapes in input (`x = np.hstack(x)`)

File: scipy/stats/morestats.py (hunk 1 above)
Comment (rgommers) [inline scipy/stats/morestats.py]: ``[[1, 2], [3, 4], [5]]`` isn't valid array-like input. I'd expect an error here. The error is quite non-descriptive, but I don't really care too much about that. I'd prefer not to add a lot of code checking for user errors like this.

If ``asarray`` would grow some kind of argument to raise an error instead of converting to object arrays, then we could get a decent error message in a concise way. But that doesn't exist yet.

Comment (rgommers) [inline scipy/stats/morestats.py]: So I'd say: just remove the above 4 lines.

**Comment by rgommers** [comment]
@aeklant can you take over https://github.com/rgommers/scipy/commit/0e4479f0fd588a2c85d624dc63614265a9af59af and https://github.com/rgommers/scipy/commit/c634a5b80f864502da997fe4df97ebca1a61a594 ?

That plus the change to remove object array handling is all that needs to be done here I think.

**Comment by aeklant** [comment]
@rgommers done.

**Comment by rgommers** [comment]
LGTM now, in it goes. Thanks @aeklant
